File: base/src/main/java/com/example/base/presenter/BasePresenter.java
package com.example.base.presenter;

import android.util.Log;

import com.example.base.view.IBaseView;
import com.example.commen.P2PError;
import com.example.commen.util.ErrorUtil;
import com.example.commen.util.LoadingPage;
import com.example.net.RetrofitCreator;
import com.google.gson.Gson;

import java.io.IOException;
import java.lang.reflect.Type;
import java.util.HashMap;
import java.util.List;

import io.reactivex.Observer;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.Disposable;
import io.reactivex.schedulers.Schedulers;
import okhttp3.ResponseBody;

/**
 * mvp的 P层抽象类,  实现网络数据的业务逻辑
 */
public abstract class BasePresenter<T> implements IBasePresenter {

    private IBaseView<T> iBaseView;

    //get请求
    @Override
    public void doHttpRequest(final int requestCode) {
        RetrofitCreator.getNetApiService().getData(getHeaderParams(), getPath(), getParams())
                .subscribeOn(Schedulers.io()) //订阅
                .observeOn(AndroidSchedulers.mainThread()) //观察
                .subscribe(new Observer<ResponseBody>() {
                    @Override
                    public void onSubscribe(Disposable d) {
                        //提示用户正在加载, 显示加载页
                        iBaseView.showLoading(LoadingPage.PAGE_LOADING_CODE);
                    }

                    @Override
                    public void onNext(ResponseBody responseBody) {
                        //隐藏加载页
                        iBaseView.hideLoading(23);

                        try {
                            String string = responseBody.string();
                            Log.i("TAG", "onNext: \n"+string);
                            //判断数据是否是列表
                            if (isList()) {
                                List<T> resEntityList = new Gson().fromJson(string, getBeanType());
                                //TODO 为判断网络数据是否请求成功, 默认成功
                                if (iBaseView != null) {
                                    iBaseView.onHttpRequestDataListSuccess(requestCode, resEntityList);
                                }
                            } else {
                                T resEntity = new Gson().fromJson(string, getBeanType());
                                if (iBaseView != null) {
                                    iBaseView.onHttpRequestDataSuccess(requestCode, resEntity);
                                }
                            }
                        } catch (IOException e) {
                            throw new RuntimeException("获取数据为空"); //扔出异常,让onError函数统一处理
                        }
                    }

                    @Override
                    public void onError(Throwable e) {
                        //获取数据失败
                        iBaseView.showLoading(LoadingPage.PAGE_ERROR_CODE);

                        String errorMessage = ErrorUtil.handleError(e);

                        //获取数据失败
                        if (iBaseView != null) {
                            iBaseView.onHttpRequestDataFailed(requestCode, P2PError.BUSINESS_ERROR);
                        }
                    }

                    @Override
                    public void onComplete() {
                    }
                });

    }


    //post请求
    @Override
    public void doHttpPostRequest(int requestCode) {

    }

    //    protected abstract Class<Object> getBeanType();
    //设置bean的类型
    protected abstract Type getBeanType();

    //设置返回的数据是否是列表
    protected boolean isList() {
        return false;
    }


    //让子类提供获取网络数据的路径
    protected abstract String getPath();

    //让子类提供调用网络请求的头参数, 例如token
    protected HashMap<String, String> getHeaderParams() {
        return new HashMap<>();
    }

    //让子类来提供调用网络请求 的参数
    protected HashMap<String, String> getParams() {
        return new HashMap<>();
    }


    @Override
    public void attachView(IBaseView iBaseView) {
        this.iBaseView = iBaseView;
    }

    @Override
    public void detachView() {
        if (iBaseView != null) {
            this.iBaseView = null;
        }
    }

}

File: base/src/main/java/com/bw/base/BasePresenter.kt
package com.bw.base

import com.bw.base.utils.ErrorUtil
import com.bw.base.utils.P2PError
import com.bw.common.AppNetConfig
import com.bw.net.ResEntity
import com.bw.net.RetrofitCreator
import com.google.gson.Gson
import io.reactivex.Observer
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.Disposable
import io.reactivex.schedulers.Schedulers
import okhttp3.ResponseBody
import java.io.IOException
import java.lang.reflect.Type


//mvp的presenter的抽象类，实现获取网络数据的业务逻辑
abstract class BasePresenter<T> : IBasePresenter<T> {

    var iBaseView:IBaseView<T>? = null

    //http get网络请求
    override fun doHttpRequest() {
        RetrofitCreator().getApiService().getMyDate(getPath())
            .subscribeOn(Schedulers.io())
            .observeOn(AndroidSchedulers.mainThread())
            .subscribe(object : Observer<ResponseBody>{
                override fun onComplete() {

                }

                override fun onSubscribe(d: Disposable) {
                    //提示用户正在加载，显示加载页
                    iBaseView?.showLoading()
                }

                override fun onNext(t: ResponseBody) {
                    //关闭显示的加载页面
                    iBaseView?.hideLoading()
                    try {
                        //如果返回的数据是列表
                        if (isList()){
                            var bean:List<T> = Gson().fromJson(t.string(),getBeanType())
                            if (iBaseView != null){
                                //获取列表数据成功
                                iBaseView?.onHttpRequestDataListSuccess(bean)
                                iBaseView?.hideLoading()
                            }
//                            else{
//                                //获取列表数据失败
//                                iBaseView?.onHttpRequestDataFailed(P2PError.BUSINESS_ERROR)
//                            }
                        }else{
                            var bean:T = Gson().fromJson(t.string(),getBeanType())
                            if (iBaseView != null){
                                //获取数据成功
                                iBaseView?.onHttpRequestDataSuccess(100,bean)
                                iBaseView?.hideLoading()
                            }
//                            else{
//                                //获取数据失败
//                                iBaseView?.onHttpRequestDataFailed(P2PError.BUSINESS_ERROR)
//                            }
                        }
                    }catch (e:IOException){
                        //抛出异常，让onError函数统一处理
                        throw RuntimeException("获取数据为空")
                        e.printStackTrace()
                    }
                }

                override fun onError(e: Throwable) {
                    val handleError = ErrorUtil.handleError(e)
                    if (iBaseView != null){
                        iBaseView?.onHttpRequestDataFailed(handleError)
                    }
//                    //关闭显示的加载页面
//                    iBaseView?.hideLoading()
//                    //获取数据失败
//                    iBaseView?.onHttpRequestDataFailed(ErrorUtil.handleError(e))
                }

            })

    }

    override fun doHttpPostRequest() {

    }

    override fun attachView(iBaseView: IBaseView<T>) {
        this.iBaseView = iBaseView
    }

    override fun datachView() {
        this.iBaseView = null
    }

    //让子类提供获取网络数据的路径
    abstract fun getPath() : String

    fun getName() : HashMap<String,String>{
        return HashMap<String,String>()
    }

    //让子类来提供调用网络请求的参数
    fun getPassWord() : HashMap<String,String>{
        return HashMap<String,String>()
    }

    fun getPhone() : HashMap<String,String>{
        return HashMap()
    }

    //让子类来提供调用网络请求的头参数，例如token
    //让子类来提供返回bean的类型
    abstract fun getBeanType() : Type

    //默认不是列表数据
    open fun isList() : Boolean{
        return false
    }

}